### PR TITLE
CLN: add typing for dtype arg in core/arrays (GH38808)

### PR DESCRIPTION
**Pull request by avinashpancham**
Follow on PR for #38808

- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/arrays/interval.py [hunk 1]
```diff
@@ -222,13 +228,14 @@ def _simple_new(
 
         if dtype is not None:
             # GH 19262: dtype must be an IntervalDtype to override inferred
-            dtype = pandas_dtype(dtype)
```
Comment (jreback) [inline pandas/core/arrays/interval.py:225]: why did you remove this convesion?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/interval.py (hunk 1 above)
Comment (avinashpancham) [reply pandas/core/arrays/interval.py:225]: Based on the logic and usage of `dtype` it seemed unnecessary. I wanted to check if all the tests will pass, if not then I will again revert it.

File: pandas/core/arrays/interval.py [hunk 2]
```diff
@@ -223,12 +229,14 @@ def _simple_new(
         if dtype is not None:
             # GH 19262: dtype must be an IntervalDtype to override inferred
             dtype = pandas_dtype(dtype)
-            if not is_interval_dtype(dtype):
+            if is_interval_dtype(dtype):
+                dtype = cast(IntervalDtype, dtype)
+                if dtype.subtype is not None:
+                    left = left.astype(dtype.subtype)
+                    right = right.astype(dtype.subtype)
+            else:
```
Comment (MarcoGorelli) [inline pandas/core/arrays/interval.py:237]: why change the logic here? If the rest of the PR just adds type hints, maybe it would be safer to keep refactorings to a separate PR?

Comment (avinashpancham) [reply pandas/core/arrays/interval.py:237]: I changed the logic because otherwise the Mypy checks will not pass.

Comment (jreback) [reply pandas/core/arrays/interval.py:237]: this is fine (you just moved and changed the logical test), but puzzled why we don't need the clause on L229

Comment (jreback) [reply pandas/core/arrays/interval.py:237]: cc @jbrockmendel if any idea here

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/interval.py [hunk 3]
```diff
@@ -212,7 +212,13 @@ def __new__(
 
     @classmethod
     def _simple_new(
-        cls, left, right, closed=None, copy=False, dtype=None, verify_integrity=True
+        cls,
+        left,
+        right,
+        closed=None,
+        copy=False,
+        dtype: Optional[Dtype] = None,
```
Comment (jbrockmendel) [inline pandas/core/arrays/interval.py:220]: by the time we get here we should have a DtypeObj shouldnt we?

File: pandas/core/arrays/interval.py [hunk 4]
```diff
@@ -223,12 +229,14 @@ def _simple_new(
         if dtype is not None:
             # GH 19262: dtype must be an IntervalDtype to override inferred
             dtype = pandas_dtype(dtype)
-            if not is_interval_dtype(dtype):
+            if is_interval_dtype(dtype):
```
Comment (jbrockmendel) [inline pandas/core/arrays/interval.py:232]: could just do `isinstance(dtype, IntervalDtype)` and avoid cast

File: pandas/core/arrays/interval.py [hunk 5]
```diff
@@ -279,7 +287,7 @@ def _simple_new(
         return result
 
     @classmethod
-    def _from_sequence(cls, scalars, *, dtype=None, copy=False):
+    def _from_sequence(cls, scalars, *, dtype: Optional[Dtype] = None, copy=False):
```
Comment (jbrockmendel) [inline pandas/core/arrays/interval.py:290]: copy: bool

File: pandas/core/arrays/interval.py [hunk 6]
```diff
@@ -338,7 +346,9 @@ def _from_factorized(cls, values, original):
             ),
         }
     )
-    def from_breaks(cls, breaks, closed="right", copy=False, dtype=None):
+    def from_breaks(
```
Comment (jbrockmendel) [inline pandas/core/arrays/interval.py:349]: copy: bool

File: pandas/core/arrays/interval.py (hunk 4 above)
Comment (avinashpancham) [reply pandas/core/arrays/interval.py:232]: In #38680, we discussed the usage of isinstance vs is_* methods in combination with cast and the latter was preferred. Therefore I continued with it here.

File: pandas/core/arrays/interval.py (hunk 3 above)
Comment (avinashpancham) [reply pandas/core/arrays/interval.py:220]: The `_simple_new` method is a.o. used in the `__new__` method on L169 of this file. The `__new__` method accepts the dtype values of type `Dtype` and passes  it without modification to the dtype arg of `_simple_new`. Therefore I also gave it type Dtype here.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @avinashpancham
